### PR TITLE
fix auto registration flake

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/44640

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/autoregistration/controller.go [hunk 1]
```diff
@@ -453,7 +454,7 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
 	// 2. connect: but the patch is based on the old workloadentry because of the propagation latency.
 	// So in this case the `DisconnectedAtAnnotation` is still there and the cleanup procedure will go on.
 	connTime := wle.Annotations[ConnectedAtAnnotation]
-	if connTime != "" {
+	if connTime != "" && c.maxConnectionAge != keepalive.Infinity {
```
Comment (ramaraochavali) [inline pilot/pkg/autoregistration/controller.go:457]: Without this, some times 1.5*maxConnAge overflows and WLE is cleaned up.

**Comment by ramaraochavali** [comment]
Tested for 500 times without failures

File: pilot/pkg/autoregistration/controller_test.go [hunk 2]
```diff
@@ -412,6 +410,24 @@ func checkEntryOrFail(
...
+	node *core.Node,
+	connectedTo string,
+	after time.Duration,
+) {
+	wait := sync.WaitGroup{}
+	wait.Add(1)
+	time.AfterFunc(after, func() {
+		wait.Done()
+	})
+	wait.Wait()
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/autoregistration/controller_test.go:436]: Was wondering if there is any diff between time.Sleep vs. time.AfterFunc

File: pilot/pkg/autoregistration/controller.go [hunk 3]
```diff
@@ -477,7 +478,7 @@ func (c *Controller) shouldCleanupEntry(wle config.Config) bool {
 	return true
 }
 
-func (c *Controller) cleanupEntry(wle config.Config) {
+func (c *Controller) cleanupEntry(wle config.Config, periodic bool) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/autoregistration/controller.go:481]: What does `periodic` do

Comment (ramaraochavali) [reply pilot/pkg/autoregistration/controller.go:481]: It was just added for logging to indicate how the WLE is cleaned up i.e is it via DelayedQueue or via Periodic cleanup

Comment (ramaraochavali) [reply pilot/pkg/autoregistration/controller.go:481]: Yes.

Comment (hzxuzhonghu) [reply pilot/pkg/autoregistration/controller.go:481]: LG

File: pilot/pkg/autoregistration/controller_test.go [hunk 4]
```diff
@@ -412,6 +409,19 @@ func checkEntryOrFail(
...
+	t test.Failer,
+	store model.ConfigStoreController,
+	wg config.Config,
+	proxy *model.Proxy,
+	node *core.Node,
+	connectedTo string,
+	after time.Duration,
+) {
+	time.Sleep(after)
+	checkEntryOrFail(t, store, wg, proxy, node, connectedTo)
```
Comment (hzxuzhonghu) [inline pilot/pkg/autoregistration/controller_test.go:422]: It always sleep that long. Why not polling periordically? I think we have some utility 
 function like wait.PollUntil

Comment (ramaraochavali) [reply pilot/pkg/autoregistration/controller_test.go:422]: This is case we have to sleep for short duration to simulate fast reconnect and check entry. It is not to poll and check

File: pilot/pkg/autoregistration/controller.go [hunk 5]
```diff
@@ -151,10 +151,12 @@ type HealthStatus = v1alpha1.IstioCondition
 // NewController create a controller which manages workload lifecycle and health status.
 func NewController(store model.ConfigStoreController, instanceID string, maxConnAge time.Duration) *Controller {
 	if features.WorkloadEntryAutoRegistration || features.WorkloadEntryHealthChecks {
-		maxConnAge := maxConnAge + maxConnAge/2
-		// if overflow, set it to max int64
-		if maxConnAge < 0 {
-			maxConnAge = time.Duration(math.MaxInt64)
+		if maxConnAge != math.MaxInt64 {
```
Comment (howardjohn) [inline pilot/pkg/autoregistration/controller.go:154]: very nit: I don't think this `if` matters, result is the same if we just always do the code?

**Review by howardjohn — APPROVED** [review]

**Pull request merged**